internal/pkg/composable/controller.go

```diff
@@ -148,43 +148,56 @@ func (c *controller) Run(ctx context.Context) error {
 
 	c.logger.Debugf("Started controller for composable inputs")
 
-	// performs debounce of notifies; accumulates them into 100 millisecond chunks
 	t := time.NewTimer(100 * time.Millisecond)
+	cleanupFn := func() {
+		c.logger.Debugf("Stopping controller for composable inputs")
+		t.Stop()
+		cancel()
+
+		// wait for all providers to stop (but its possible they still send notifications over notify
```

Comment:
FYI there is actually another deadlock where the agent will wait for the providers to exit forever: https://github.com/elastic/elastic-agent/issues/2224